packages/playwright-core/src/server/trace/recorder/tracing.ts

```diff
@@ -570,18 +529,90 @@ function createAfterActionTraceEvent(metadata: CallMetadata): trace.AfterActionT
-  if (event.type === 'event' || event.type === 'object')
-    return;
+  writeFile(file: string, content: string | Buffer, skipIfExists?: boolean) {
+    this._buffers.delete(file); // No need to flush the buffer since we'll overwrite anyway.
 
-  await flushTraceFile(file);
-}
+    // Note: 'wx' flag only writes when the file does not exist.
+    // See https://nodejs.org/api/fs.html#file-system-flags.
+    // This way tracing never have to write the same resource twice.
+    this._appendOperation(async () => {
+      if (skipIfExists)
+        await fs.promises.writeFile(file, content, { flag: 'wx' }).catch(() => {});
+      else
+        await fs.promises.writeFile(file, content);
+    });
+  }
+
+  appendFile(file: string, text: string, flush?: boolean) {
+    if (!this._buffers.has(file))
+      this._buffers.set(file, []);
+    this._buffers.get(file)!.push(text);
+    if (flush)
+      this._flushFile(file);
+  }
 
-async function flushTraceFile(file: { file: string, buffer: trace.TraceEvent[] }) {
-  const data = file.buffer.map(e => Buffer.from(JSON.stringify(e) + '\n'));
-  await fs.promises.appendFile(file.file, Buffer.concat(data));
-  file.buffer = [];
+  private _flushFile(file: string) {
+    const buffer = this._buffers.get(file);
+    if (buffer === undefined)
+      return;
+    const text = buffer.join('');
+    this._buffers.delete(file);
+    this._appendOperation(() => fs.promises.appendFile(file, text));
+  }
+
+  copyFile(from: string, to: string) {
+    this._flushFile(from);
+    this._buffers.delete(to); // No need to flush the buffer since we'll overwrite anyway.
+    this._appendOperation(() => fs.promises.copyFile(from, to));
+  }
+
+  async syncAndTakeError() {
+    for (const file of this._buffers.keys())
+      this._flushFile(file);
+    await this._writeChain;
+    const result = this._error;
+    this._error = undefined;
+    return result;
+  }
+
+  zip(entries: NameValue[], zipFileName: string) {
+    for (const file of this._buffers.keys())
+      this._flushFile(file);
+
+    // Chain the export operation against write operations,
+    // so that files do not change during the export.
+    this._appendOperation(async () => {
+      const zipFile = new yazl.ZipFile();
+      const result = new ManualPromise<void>();
+      (zipFile as any as EventEmitter).on('error', error => result.reject(error));
+      for (const entry of entries)
+        zipFile.addFile(entry.value, entry.name);
+      zipFile.end();
+      zipFile.outputStream
+          .pipe(fs.createWriteStream(zipFileName))
+          .on('close', () => result.resolve())
+          .on('error', error => result.reject(error));
+      await result;
+    });
+  }
+
+  private _appendOperation(cb: () => Promise<unknown>): void {
+    // This method serializes all writes to the trace.
+    this._writeChain = this._writeChain.then(async () => {
+      try {
+        await cb();
+      } catch (e) {
+        this._error = e;
```

Comment:
Do you want to stop the chain in case of the error? Otherwise all these operations will be happening and you would only capture the last error.